hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+# Versioning
+
+Specification for how storage versioning will operate for code generated CRD definitions.
+
+## Goals
+
+**Auto-generated conversions:** As far as practical, we want to autogenerate the schema for each hub version along with the conversions to and from the actual ARM API versions. Hand coding all the required conversions doesn't scale across all the different Azure sevices, especially with the ongoing rate of change. 
+
+**Allow for hand coded conversions:** While we expect to use code generation to handle the vast majority of needed conversions, we anticipate that some breaking API changes will require some of the conversion to be hand coded. We need to make it simple for these conversions to be introduced while still automatically generating the majority of the conversion.
+
+**No modification of generated files.** Manual modification of generated files is a known antipattern that greatly increases the complexity and burden of updates. If some files have been manually changed, every difference showing after code generation needs to be manually reviewed before being committed. This is tedious and error prone because the vast majority of auto generated changes will be perfectly fine. 
+
+**Compliance with Kubernetes versioning.** To quote [Kubebuilder's documentation](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html):
+
+> In Kubernetes, all versions must be safely round-tripable through each other. This means that if we convert from version 1 to version 2, and then back to version 1, we must not lose information. Thus, any change we make to our API must be compatible with whatever we supported in v1, and also need to make sure anything we add in v2 is supported in v1.
+
+## Non-Goals
+
+**Coverage of every case by code generation:** While it's likely that very high coverage will be achievable with code generation, we don't believe that it will be practical to handle every possible situation automatically. It's therefore necessary for the solution to have some form of extensibility allowing for the injection of hand written code.
+
+## Proposed Solution
+
+In summary (details below):
+
+* Base the storage version on the latest available API version for each resource type
+* Augment the storage version with a name/value property bag for storing additional information
+* Generate the required `ConvertTo()` and `ConvertFrom()` methods automatically
+* Allow configuration to support common changes from version to version
+* Define extension points for hand written conversions
+
+### Use the latest available API Version
```

Comment:
We aren't actually using the latest API version are we -- we're just using it _as a base_?

So for example if we have 3 API versions currently: `v1`, `v2`, `v3`, then we would also have a `v3storage` basically, right? Or is `v3` == `v3Storage`?

If we're just using it as a base we need to talk about what happens when the base changes. How do we get from `v3storage` to `v4storage`

Reply:
I'm confused about this. If a new version comes out and we regenerate the storage type and conversion functions, will the storage type be updated to match the latest API version type? 

Won't that mean that running controllers that wrote a `v2storage` into the k8s API and are then upgraded will try to read a `v3storage`? 

It seems like the storage types need to be versioned independently from the API types. Maybe that's ok - old storage types just become extra spokes for the new hub? Not sure whether the kubebuilder conversion stuff will handle that though.

Reply:
Correct, if we have 3 API versions currently (`v1`, `v2` & `v3`) then we'll create a hub/storage version based on `v3` called (say) `vStorage`. 

We do need to think through the upgrade scenario. The ASO repo uses a webhook to trigger the upgrade when you deploy a new version of the operator; I'm not sure of the exact details.

Reply:
Ah, ok - so there's a moment when the operator can load the old versions of specs and convert them to the new version?

I guess the key thing is that if `vStorage` is based on `v3` because that was the latest API version when `vStorage` was created, we don't necessarily need to update `vStorage` when `v4` comes out (unless there's something that can't be represented).

Reply:
I think we would *always* update to v4 (the latest version), right?

Reply:
Is there a win in having `v3` and `v3Storage` be different packages? How are they actually different? How is the generation of `ConvertTo`'s work when we come out with `v4` and are making `v4Storage` the storage shcema? It feels like we need a `v3` -> `v4Storage` and a `v3Storage` -> `v4Storage` and it's not clear to me how they will ever differ.

Reply:
I think I've covered this off in the new spec, let me know if it reads clearly.